### PR TITLE
Change authorized applications page

### PR DESCRIPTION
**Pull request by Gargron**
Start tracking when an OAuth access token was last used and from which IP (retain IP data only as long as our general IP retention period allows). Change the design of the authorized applications page in account settings to be more helpful. Change how application permissions are presented.

image

image

### CODE REVIEW

File: app/views/oauth/authorized_applications/index.html.haml [hunk 1]
```diff
@@ -1,24 +1,43 @@
...
+.announcements-list
+  - @applications.each do |application|
+    .announcements-list__item
+      - if application.website.present?
+        = link_to application.name, application.website, target: '_blank', rel: 'noopener noreferrer', class: 'announcements-list__item__title'
+      - else
+        %strong.announcements-list__item__title
+          = application.name
+          - if application.superapp?
+            %span.account-role.moderator= t('doorkeeper.authorized_applications.index.superapp')
```
Comment (ClearlyClaire) [inline app/views/oauth/authorized_applications/index.html.haml:17]: Not new but we probably shouldn't repurpose all those classes… either rename them to something more neutral or have the styles apply to different classes.

File: app/javascript/styles/mastodon/containers.scss [hunk 2]
```diff
@@ -110,7 +107,7 @@
     display: block;
     font-size: 32px;
     line-height: 40px;
-    margin-left: 8px;
+    margin-left: 10px;
```
Comment (ClearlyClaire) [inline app/javascript/styles/mastodon/containers.scss:110]: All those changes seem unrelated?

File: db/schema.rb [hunk 3]
```diff
@@ -1231,4 +1210,27 @@
...
+              session_activations.updated_at
+             FROM session_activations
+          UNION ALL
+           SELECT login_activities.user_id,
+              login_activities.ip,
+              login_activities.created_at
+             FROM login_activities
+            WHERE (login_activities.success = true)) t0
+    GROUP BY t0.user_id, t0.ip;
+  SQL
```
Comment (ClearlyClaire) [inline db/schema.rb:1235]: Not necessarily an issue, but I don't understand why this has moved around.

File: app/helpers/application_helper.rb [hunk 4]
```diff
@@ -224,4 +224,19 @@ def render_initial_state
     content_tag(:script, json_escape(json).html_safe, id: 'initial-state', type: 'application/json')
     # rubocop:enable Rails/OutputSafety
   end
+
+  def grouped_scopes(scopes)
+    scope_parser      = ScopeParser.new
+    scope_transformer = ScopeTransformer.new
+
+    scopes.each_with_object({}) do |str, h|
+      scope = scope_transformer.apply(scope_parser.parse(str))
```
Comment (ClearlyClaire) [inline app/helpers/application_helper.rb:233]: Surely the scopes already do get parsed and you don't need to roll your own parser?

**Review by ClearlyClaire — COMMENTED** [review]
Not a full review as I have not read the parsing code yet, but I like the idea and the redesign, although I have some comments on the design itself:
- I'm afraid it may take a lot of space per app, but I guess it's manageable
- I think it would be better if there was a clearer visual separation between apps. They don't need to be in the same table.

File: app/views/oauth/authorized_applications/index.html.haml [hunk 5]
```diff
@@ -1,24 +1,43 @@
...
+            = t('doorkeeper.authorized_applications.index.last_used_at', date: l(application.most_recently_used_access_token.last_used_at.to_date))
+          - else
+            = t('doorkeeper.authorized_applications.index.never_used')
+
+          •
+
+          = t('doorkeeper.authorized_applications.index.authorized_at', date: l(application.created_at.to_date))
+
+        %div
+          = table_link_to 'times', t('doorkeeper.authorized_applications.buttons.revoke'), oauth_authorized_application_path(application), method: :delete, data: { confirm: t('doorkeeper.authorized_applications.confirmations.revoke') }
```
[the remark this answers is not in the document]
Comment (Gargron) [reply app/views/oauth/authorized_applications/index.html.haml:31]: It’s possible for a user to accidentally leak the web app’s access token, so it would be nice to have a reset button. The web app will simply generate a new token on next access.

File: db/schema.rb (hunk 3 above)
Comment (Gargron) [reply db/schema.rb:1235]: This is my development environment’s database. I’ll reset.

File: app/helpers/application_helper.rb (hunk 4 above)
Comment (Gargron) [reply app/helpers/application_helper.rb:233]: No, they do not get parsed like I want. They are already an array of strings, but that’s not enough. We have a system for designating access type in the way our scopes are named. This allows me to group them better.

File: app/javascript/styles/mastodon/containers.scss (hunk 2 above)
Comment (Gargron) [reply app/javascript/styles/mastodon/containers.scss:110]: The account header is displayed above the OAuth prompt. I’m making sure it looks good together.

File: app/views/oauth/authorized_applications/index.html.haml (hunk 5 above)
[the remark this answers is not in the document]
Comment (ClearlyClaire) [reply app/views/oauth/authorized_applications/index.html.haml:31]: I tried clicking the link, it causes subsequent accesses to the WebUI to fail every request with “401 The access token was revoked”, including with a page refresh. This requires logging out and logging in again. So if we do offer this, it probably should renew the token for the current session as well.

**Review by ClearlyClaire — APPROVED** [review]
Design-wise, I'd largely prefer if applications were each a different table / at least more visually separated.

Code-wise, I don't like that we are repurposing CSS classes which are now ill-named, but I can't hold that against this PR specifically, this is something that'll have to be addressed separately.

Otherwise this looks good to me!

File: app/helpers/application_helper.rb (hunk 4 above)
Comment (ClearlyClaire) [reply app/helpers/application_helper.rb:233]: I wonder if involving parslet isn't a little overkill, but it looks sound, and I understand why it is here.

**Comment by tribela** [comment]
Last used timestamp would very helpful if user want to get rid of old sessions
👍🏼
